src/main/java/com/eternalcode/parcellockers/deliverycode/repository/DeliveryCodeRepositoryImpl.java

```diff
@@ -0,0 +1,86 @@
+import com.eternalcode.parcellockers.deliverycode.DeliveryCode;
+
+import javax.sql.DataSource;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+
+public class DeliveryCodeRepositoryImpl extends AbstractDatabaseService implements DeliveryCodeRepository {
+
+    private final Map<UUID, DeliveryCode> cache = new HashMap<>();
+
+    protected DeliveryCodeRepositoryImpl(DataSource dataSource) {
+        super(dataSource);
+
+        this.initTable();
+    }
+
+    private void initTable() {
+        this.executeSync("CREATE TABLE IF NOT EXISTS `delivery_codes`(" +
+            "`parcelUUID` VARCHAR(36) NOT NULL, " +
+            "`code` VARCHAR(6) NOT NULL, " +
+            "PRIMARY KEY (parcelUUID) " +
+            ");", PreparedStatement::execute);
+    }
+
+    @Override
+    public CompletableFuture<Void> save(DeliveryCode deliveryCode) {
+        return this.execute("INSERT INTO `delivery_codes` (`parcelUUID`, `code`) VALUES (?, ?);", preparedStatement -> {
+            preparedStatement.setString(1, deliveryCode.parcelUUID().toString());
+            preparedStatement.setString(2, deliveryCode.code());
+            preparedStatement.execute();
+            this.cache.put(deliveryCode.parcelUUID(), deliveryCode);
+        });
+    }
+
+    @Override
+    public CompletableFuture<DeliveryCode> findByUUID(UUID parcelUUID) {
+        return this.supplyExecute("SELECT * FROM `delivery_codes` WHERE `parcelUUID` = ?;", statement -> {
+            statement.setString(1, parcelUUID.toString());
+            ResultSet rs = statement.executeQuery();
+
+            if (!rs.next()) {
+                return null;
+            }
+
+            this.cache.putIfAbsent(parcelUUID, new DeliveryCode(
+                UUID.fromString(rs.getString("parcelUUID")),
+                rs.getString("code")
+            ));
+
+            return new DeliveryCode(
+                UUID.fromString(rs.getString("parcelUUID")),
+                rs.getString("code")
+            );
+        });
+    }
+
+    @Override
+    public CompletableFuture<Void> remove(DeliveryCode deliveryCode) {
+        return this.remove(deliveryCode.parcelUUID());
+    }
+
+    @Override
+    public CompletableFuture<Void> remove(UUID parcelUUID) {
+        return this.supplyExecute("DELETE FROM `delivery_codes` WHERE `parcelUUID` = ?;", statement -> {
+            statement.setString(1, parcelUUID.toString());
+            statement.execute();
+            this.cache.remove(parcelUUID);
+            return null;
+        });
+    }
+
+    public Optional<DeliveryCode> find(UUID parcelUUID) {
+        if (this.cache.get(parcelUUID) != null) {
+            return Optional.of(this.cache.get(parcelUUID));
+        }
+        return Optional.ofNullable(this.findByUUID(parcelUUID).join());
```

Comment:
```suggestion
       Optional<DeliveryCode> deliveryCodeOptional = Optional.ofNullable(this.cache.get(parcelUUID));
       if (deliveryCodeOptional.isPresent()) {
           return deliveryCodeOptional;    
       }

        return this.findByUUID(parcelUUID).join();
```

Reply:
ten optional słabo wygląda

Reply:
To mam revertować te optionale?

Reply:
To można zrobić jeszcze:
`Optional.ofNullable(this.cache.get(parcelUUID)).or(() -> this.findByUUID(parcelUUID).join));`

Reply:
**sorry, że piszę po polsku zapomniałem, że nie wolno. 

1.
PL: Ogólnie to pakowanie w optionala w tym samym scope i go sprawdzanie mija się z celem np.:
EN: Packing an optional in the same scope and checking it is pointless
```java
Optional optional = Optional.ofNullable(getNullable());

if (optional.isEmpty()) {
    // ....
}
```

2.
PL: to co zrobił DMK jest już dobre
EN: what DMK has done is already good
```java
return Optional.ofNullable(this.cache.get(parcelUUID)).or(() -> this.findByUUID(parcelUUID).join));
```

3,
PL: Robienie tutaj join mija się z celem i jest też kompletnie bez kontekstu (bez use case)
EN: `join()` here is pointless and is also completely without context (without use case)

4. 
EN: Cache should only be in the **worst** possible case. please remove it